Add mocking prototype

JulianHarrison-Wirth opened the pull request:
## Main changes of this PR

Adds a **mock of the preCICE Participant API** in `extras/mock/`, so adapters can be developed and tested without running real preCICE and without a coupling partner.

* **`libpreciceMocked.so`** (CMake target `preciceMocked`): a drop-in replacement for `libprecice.so` that re-implements the complete `precice::Participant` API in `extras/mock/preciceMocked.cpp`. It does not link against `libprecice` and only needs a subset of its dependencies — a C++17 compiler, Eigen, libxml2, the bundled fmt, and optionally MPI. No Boost, no PETSc, no communication backends.
* **Two ways to use it**: preload it into an unmodified solver (`LD_PRELOAD=/path/to/libpreciceMocked.so ./run.sh`, which also covers solvers using the language bindings), or link `preciceMocked` instead of `precice`.
* **Configuration-aware validation**: the mock reads your regular preCICE configuration and validates every API call against it — participant, mesh and data names, read/write direction, buffer sizes, and the legal order of steering calls — and reports violations as `precice::Error` with an actionable message. Time stepping, subcycling, implicit iterations and checkpoint requirements are derived from the configured coupling scheme.
* **Optional `precice-mock-config.xml`** (placed next to the preCICE configuration) controls what `readData()` returns — `buffer` (echoes what was written, default), `scaled` (scalar or element-wise vector multiplier), or `random` (with bounds and seed) — per mesh/data pair or globally, plus the logging mode and a `max-iterations` override for implicit coupling.
* **C and Fortran bindings** are compiled into the mock library as well. To make that possible they no longer depend on preCICE internals (`logging/Logger`, `logging/LogMacros`, `utils/assertion`): the `PRECICE_CHECK` guards for "participant not created (yet)" now print to `std::cerr` and `std::abort()`, which matches the previous observable behaviour without pulling in Boost.Log.
* **Tests** (`extras/mock/tests/`, see `extras/mock/tests/README.md`):
  * *Golden API tests* — 26 ctest cases (`ctest -L mock`). A driver (`tests/api/driver.cpp`) exercises the API scenario by scenario (explicit, implicit, subcycling, max-time, first-participant, error paths) and its output is diffed against committed expectations. Since the mock is deterministic, any diff is a behaviour change.
  * *Differential harness (manual)* — `tests/api/run.sh` runs the same driver against the real `libprecice` as a coupled pair and against the mock, and diffs both. This is how the golden files are (re)validated whenever real preCICE behaviour changes.
  * *Tutorial smoke tests (manual, ~4.5 min)* — `tests/tutorials/smoke.sh` runs real adapters (OpenFOAM, python bindings, FMI, nutils, C++) from a curated set of tutorials plus the four preCICE course solution sets against the mock, and compares normalized output against committed baselines. `smoke_update.sh` regenerates those baselines, first validating every case against real preCICE.
* **Docs**: `extras/mock/README.md` (building, usage, configuration, behaviour, limitations) and `extras/mock/tests/README.md`.
* `.pre-commit-config.yaml`: excludes the committed tutorial baselines from codespell.

### Known limitations (documented in the README)

* **No real communication**: each rank mocks the coupling independently and reads back rank-local buffers. An MPI communicator passed to the constructor is validated but not used, so `getMeshVertexSize()` on a received mesh without API access returns 0.
* **No convergence measures**: implicit schemes converge after a fixed number of iterations (default 2, overridable) instead of evaluating the configured measures.
* **Connectivity and gradient data** are validated but not stored (there is no read counterpart for gradients in the API).
* If a configuration bounds neither `max-time` nor `max-time-windows`, the mock adds `max-time-windows="100"` and warns — valid for real preCICE, but the mock has no partner that could end the simulation.

## Motivation and additional information

Testing an adapter today requires either a full preCICE installation plus a second participant, or a hand-written mock maintained inside the adapter (as the Python bindings had in `test/Participant.cpp`). Both are slow to set up, awkward in CI, and expose tests to failures outside the code under test.

A single mock maintained next to preCICE gives adapter developers a fast, dependency-light way to exercise API usage, error paths and checkpoint handling, and lets bindings and adapters drop their private mocks. A companion PR for the Python bindings (precice/python-bindings#XXX) does exactly that: it removes the hand-written mock and runs the full test suite against `libpreciceMocked.so`, which also grew the suite from 65 to 76 tests, since the mock allows testing behaviour the old stub could not express.

This is a first complete-but-simple implementation and deliberately does not try to be a preCICE simulator — feedback on scope and on the mock configuration format is very welcome.

## Author's checklist

* [x] I used the [`pre-commit` hook](https://precice.org/dev-docs-dev-tooling.html#setting-up-pre-commit) to prevent dirty commits and used `pre-commit run --all` to format old commits.
* [ ] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [x] I added a test to cover the proposed changes in our test suite.
* [x] For breaking changes: I documented the changes in the appropriate [porting guide](https://precice.org/couple-your-code-porting-overview.html).
* [x] I stuck to C++17 features.
* [x] I stuck to CMake version 3.22.1.
* [ ] I squashed / am about to squash all commits that should be seen as one.
* [ ] I ran the systemtests by adding the label `trigger-system-tests` (may be skipped if minor change)

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?
* [ ] Is `extras/mock/` the right place for this, and should `preciceMocked` be installed/exported or stay a build-tree-only target?
* [ ] Are the C/Fortran bindings acceptable without preCICE's logging and assertion headers (`std::cerr` + `std::abort()` instead of `PRECICE_CHECK`)?
* [ ] Is the `precice-mock-config.xml` format (modes, defaults, iteration override) a good long-term interface for adapter developers?
* [ ] Are the committed tutorial baselines (~75k lines) acceptable in this repository, or should the smoke tests live elsewhere?
* [ ] Should the golden API tests (`ctest -L mock`) run in CI by default?

fsimonis committed:
Remove logger usage in bindings

fsimonis committed:
Add preciceMocked

JulianHarrison-Wirth committed:
Initial No-Op mock

JulianHarrison-Wirth committed:
Merge branch 'develop' into add-mocking-prototype

JulianHarrison-Wirth committed:
Fixed compilation issues
Added Data Outputting from file
Updated .gitignore

JulianHarrison-Wirth committed:
ran pre-commit

JulianHarrison-Wirth committed:
added personal todos

JulianHarrison-Wirth committed:
Merge remote-tracking branch 'Origin/develop' into add-mocking-prototype

JulianHarrison-Wirth committed:
Added MPI communication check to compilation of mocked interface to fix mpi error on compilation

JulianHarrison-Wirth committed:
Added assertions

fsimonis commented:
@MakisH I just realized that I converted the checks straight to if conditions. So they are all flipped...

fsimonis committed:
Fix conditions in bindings

JulianHarrison-Wirth committed:
Added Comments

JulianHarrison-Wirth committed:
ran pre-commit

JulianHarrison-Wirth committed:
Testing

JulianHarrison-Wirth committed:
Added correct error handling akin to precice using the precice error library

added support for implicit coupling

added the ability to read precice configs and check for errors, when the adapter doesn't do what the config expects

added a mock-config (+example) to choose data mode

added script to test all tutorials

added basic readme

fixed shit

JulianHarrison-Wirth committed:
undid testing changes

JulianHarrison-Wirth commented:
correction: I got 40 passes
all of the ones requiring pyprecice in a venv failed due to the newest version of pyprecice throwing an error on install
for manual install I was able to use version 3.2.1 but I don't know how that works with venv
the error was: Failed building wheel for pyprecice
and It said something about a missing/wrong licence

MakisH committed:
Merge branch 'develop' into add-mocking-prototype

MakisH committed:
Merge branch 'mocked-api' into add-mocking-prototype

MakisH commented on examples/solverdummies/cpp/CMakeLists.txt:
```diff
@@ -7,7 +7,6 @@ add_executable(solverdummy solverdummy.cpp)
 set_target_properties(solverdummy PROPERTIES CXX_STANDARD 11)
 target_link_libraries(solverdummy PRIVATE precice::precice)
 
-
```
Left-over from debugging / unrelated change.

MakisH commented on extras/mock/README.md:
```diff
@@ -0,0 +1,172 @@
+## Testing Your Adapter
+
+1. Create a simple preCICE config with your participant
+2. Optionally create a mock-config.xml for specific data behavior
```
```suggestion
2. Optionally, create a `mock-config.xml` for specific data behavior
```

JulianHarrison-Wirth replied:
added

MakisH commented on extras/mock/preciceMocked.cpp:
```diff
@@ -0,0 +1,1450 @@
+    if (solverProcessIndex >= solverProcessSize) {
+      throw precice::Error(precice::utils::format_or_error("solverProcessIndex={} must be smaller than solverProcessSize={}.", solverProcessIndex, solverProcessSize));
+    }
+    seed = static_cast<uint32_t>(rank) ^ 0x9e3779b9u;
```
What is this for? Add a comment.

MakisH commented on extras/mock/preciceMocked.cpp:
```diff
@@ -0,0 +1,1450 @@
+
+  // std::unique_ptr<FileDoubleReader> fileReader;
+
+  void parseConfig()
```
I wonder if we could simplify the parsing, reusing existing implementations. Maintaining this will be complicated once the config format changes again.

On the other hand, I see that this is only extracting minimal information.

MakisH reviewed (commented):
Some first comments, mostly on the high level so far. I have only gone through ~half the `preciceMocked.cpp`. Looks very nice so far!

Note that, to make the review easier, I cherry-picked the skeleton-related commits into the `mocked-api` branch, merged `develop` into both branches (since you had already merged as well), and merged `mocked-api` to your branch, resolving a conflict in `CMakeLists.txt`. The skeleton-related commits are still shown here, but the `files changed` view is clearer. Feel free to rebase later.

JulianHarrison-Wirth committed:
verschlimmbessert
fails more examples but is more realistic

JulianHarrison-Wirth committed:
Merge branch 'add-mocking-prototype' of https://github.com/JulianHarrison-Wirth/precice into add-mocking-prototype

Merged local changes with remote

JulianHarrison-Wirth commented on extras/mock/README.md:
```diff
@@ -0,0 +1,172 @@
+
+## Configuration Files
+
+### preCICE Configuration
```
Implementing would not be a huge effort, but since most callers will have the config anyway I am not sure if it is necessary 
I can add it as an option later if wanted

uekerman replied:
I see the point, but I don't think we need such a feature.
While developing an adapter, I want to mainly check whether I call set_mesh_vertices or read/write_data correctly. For all of these, I need names.

We already now have enough tooling to easily generate a first config.

And developing an adapter without targeting a specific application is nothing we have seen so far.

JulianHarrison-Wirth commented on extras/mock/README.md:
````diff
@@ -0,0 +1,172 @@
+target_link_libraries(your_solver preciceMocked)
+```
+
+Alternatively build normally and use LD_PRELOAD to load the mock at runtime
````
swaped the order

JulianHarrison-Wirth commented on extras/mock/run_all_tutorials.sh:
that was just a quick benchmark for dev purposes to see if I was breaking stuff

JulianHarrison-Wirth commented on extras/mock/mock-config.xml.example:
pre-commit didnt let me have comments in an xml file

JulianHarrison-Wirth commented on extras/mock/preciceMocked.cpp:
```diff
@@ -0,0 +1,1450 @@
+  std::map<std::string, std::vector<double>> writeBuffers;
+
+  // std::unique_ptr<FileDoubleReader> fileReader;
+
```
should be gone now

JulianHarrison-Wirth committed:
Small Changes

added changes from comments
added a test for courses
removed unnecessary example file
updated readme

JulianHarrison-Wirth commented on extras/mock/preciceMocked.cpp:
```diff
@@ -0,0 +1,1635 @@
+  _impl->parseMockConfig();
+
+  // Communication setup prints
+  if (_impl->rank == 0) {
```
what would the advantage of abstracting it to a new member variable  (primary = (rank == 0))?

fsimonis replied:
Less to type and easier to follow. We also do this in the main library.

JulianHarrison-Wirth commented on extras/mock/preciceMocked.cpp:
```diff
@@ -0,0 +1,1635 @@
+  _impl->parseConfig();
+
+  // Parse mock configuration
+  _impl->parseMockConfig();
```
removed duplicate calls, parseConfig() is still guarded against multiple calls

JulianHarrison-Wirth commented on extras/mock/preciceMocked.cpp:
```diff
@@ -0,0 +1,1635 @@
+// check if vector components are written correctly (later/put in issues)
+// --read config (parts) to see if adapter does what config expects from it !!
+// --names, dims...
+// --libxml2 wrapper in utils or xml !! Not feasible due to requiring linking the full precice library
```
I have implemented it using a SAX handler now

fsimonis replied:
Very nice

JulianHarrison-Wirth committed:
First Review Changes

refactored the mock following suggestions
some small changes suggested in the reviews
fixed some wrong behavior regarding 2D meshes and other cases not correctly accounted for
fixed errors introduced by refactoring
did not implement std::span or std::string view because it changes the ABI and causes LD_PRELOAD to fail

now correctly errors on the courses (same courses fail as with real precice)
passes 45/92 tutorials, no failures are due to wrong implementations of the mock (8 are crashes/errors of the solver due to mocked data causing float errors or similar, 13 are due to pip install pyprecice erroring on my system in a venv and the rest are due to missing executables, but as this is not my goal to optimize and was just a dev benchmark for myself I will not further persue this avenue to reduce the amount of fails)

updated CMakeLists.txt to use C++20 for include format
removed stupid mistake from run_course_soulution.sh

JulianHarrison-Wirth committed:
actually commited the right file

JulianHarrison-Wirth committed:
updated to be fully threadsafe

JulianHarrison-Wirth committed:
Added consistent Mutexes
replaced format with fmt to go backk to c++ 17

JulianHarrison-Wirth committed:
add illustrations

JulianHarrison-Wirth committed:
fixedd potential deadlock

JulianHarrison-Wirth committed:
temp fix

JulianHarrison-Wirth committed:
removed impl checks and made mutexes correctly recursive

JulianHarrison-Wirth committed:
Added global default mode to config

JulianHarrison-Wirth committed:
added stricter validation to data reads

JulianHarrison-Wirth committed:
fixed bugs and added some verbosity
added iteration limit to mock to prevent instant end of cycling for implicit coupling
made buffer universal and added handling for missmatched sizes without the buffer
added warning if scaling factor is given in buffered mode or no scaling factor is given in scaled mode
replaced some missed std::cerr with precice errors

JulianHarrison-Wirth committed:
added bounds to prng

JulianHarrison-Wirth committed:
made xml consistent

JulianHarrison-Wirth committed:
fixed bugs and added some verbosity

added iteration limit to mock to prevent instant end of cycling for implicit coupling

made buffer universal and added handling for missmatched sizes without the buffer

added warning if scaling factor is given in buffered mode or no scaling factor is given in scaled mode

replaced some missed std::cerr with precice errors

Added bounds to prng

JulianHarrison-Wirth committed:
Merge branch 'add-mocking-prototype' of https://github.com/JulianHarrison-Wirth/precice into add-mocking-prototype

JulianHarrison-Wirth committed:
added seed to prng xml

JulianHarrison-Wirth committed:
fixed build issue

JulianHarrison-Wirth committed:
fixed implicit termination

JulianHarrison-Wirth committed:
hopefully really fixed implicit behavior

MakisH commented on extras/mock/mock-config.xml:
Hopefully trivial change, but useful for clarity: let's rename `mock-config.xml` to `precice-mock-config.xml`.

This is for consistency with all other output files, for example.

JulianHarrison-Wirth committed:
hopefully fixed mac building errors

JulianHarrison-Wirth committed:
cleaned up naming and comments

JulianHarrison-Wirth committed:
Merge branch 'add-mocking-prototype' of https://github.com/JulianHarrison-Wirth/precice into add-mocking-prototype

JulianHarrison-Wirth committed:
Revert "Merge branch 'add-mocking-prototype' of https://github.com/JulianHarrison-Wirth/precice into add-mocking-prototype"

This reverts commit cb1ec861ec0c11cda882a7289cd8bdd8db26b4e5, reversing
changes made to 7984e959d9717de4fe842f75f0807bbf0a7df403.

JulianHarrison-Wirth commented:
@fsimonis If you could look at the build issues on mac please, it seems to be a cmake issue

JulianHarrison-Wirth committed:
fixed typo and adjusted default behavior

JulianHarrison-Wirth committed:
moved 2 files

JulianHarrison-Wirth committed:
adjusted testin script

JulianHarrison-Wirth committed:
changed default mocked data to mocked data default

JulianHarrison-Wirth committed:
removed dev scripts

uekerman reviewed (commented):
Some quick comments already. I still want to try out the mock actually.

uekerman commented on extras/mock/README.md:
```diff
@@ -0,0 +1,286 @@
+- `max-time-windows`: Terminates after completing the specified number of time windows (each time window may involve multiple sub-iterations)
+- `max-time`: Terminates at the end of a time window when the accumulated time reaches or exceeds the limit (does not terminate during sub-cycling)
+
+Both coupling schemes will throw an error if neither termination criterion is provided.
```
I think that all of this is not needed. This is normal preCICE behavior.

Maybe one overall h2 section "Configuration" could be a better structure?

uekerman commented on extras/mock/README.md:
```diff
@@ -0,0 +1,286 @@
+### Mock Configuration (Optional)
+Place `precice-mock-config.xml` in the same directory as your preCICE config file.
+
+**Example precice-mock-config.xml:**
```
The config is also included here. No need to duplicate. We only have to document where to find the example.

uekerman commented on extras/mock/README.md:
```diff
@@ -0,0 +1,286 @@
+- Returns `true` from `requiresWritingCheckpoint()` at the start of each iteration
+- Returns `true` from `requiresReadingCheckpoint()` when iterating (not converged)
+- Tracks iteration count and convergence status
+- Reports time window completion correctly
```
That's all expected behavior, does not need documentation IMO.

uekerman commented on extras/mock/README.md:
```diff
@@ -0,0 +1,286 @@
+## Limitations
+
+- Does not perform actual mesh mapping or data interpolation
+- Does not communicate between participants (single-process only)
```
"between participants" will lead to confusions.
"single-process only" is also not true, right? You have tried to run with a parallel participant. Let's maybe try to boil it down here to the actual limitation (cannot handle MPI communicator in constructor? or maybe there is none in  the end?).

uekerman commented on CMakeLists.txt:
```diff
@@ -576,6 +576,18 @@ target_include_directories(preciceMocked PUBLIC
   $<INSTALL_INTERFACE:include>
   )
 
+target_link_libraries(preciceMocked PUBLIC Eigen3::Eigen)
```
Should we document the dependencies of the mock somewhere. All are dependencies of preCICE as well, but it's a proper subset. You could link to the dependencies of preCICE and then state which ones are needed.

uekerman commented on extras/mock/preciceMocked.cpp:
```diff
@@ -0,0 +1,1796 @@
+    throw precice::Error("requiresInitialData() has to be called before initialize().");
+  }
+  return false;
+}
```
This could be sth to read from the preCICE config. Otherwise, document as limitation.

uekerman reviewed (changes requested):
I had closer look from the user perspective. I did not look at the implementation of the configuration.

Besides the points below:
- Let's try to condense the documentation to the important things.
- We are building the mock as part of the CI already, right? Could be good to also add some smoke test (e.g. solver dummies)
- As mentioned at the presentation: An interesting test would be to run all preCICE integration tests and see what happens (crash vs. expected errors vs. unexpected errors). For example, the missing support for subcycling should be caught like this. 
- To what extend does the mock support [remeshing](https://precice.org/couple-your-code-moving-or-changing-meshes.html#remeshing-using-precice)?

JulianHarrison-Wirth committed:
Adressed many comments:
fixed timestep remaining time
added order check for profiling sections
add mock version information and enhance logging modes in mock participant

uekerman commented:
@JulianHarrison-Wirth Thank you for the update. As this is a huge PR, which becomes hard to keep an overview of. To mitigate, I suggest resolving "done" comments and keeping commits small: one commit per non-trivial change.

In the current state, I don't believe that subcycling already works. The handling in `advance` still seems wrong. You should be able to easily test this with a solverdummy that subcycles. Maybe unclear: With subcycling we simply mean that a participant uses a time step size smaller than the time window size. More information in the [docs on time stepping](https://precice.org/couple-your-code-time-step-sizes.html). Please ask for help if unclear.

JulianHarrison-Wirth committed:
fixed small logging order issue

JulianHarrison-Wirth commented on extras/mock/README.md:
```diff
@@ -0,0 +1,276 @@
+# preCICE Mock Participant
```
do you need me to do anything here or is this just waiting for the final merge?

MakisH replied:
The file mainly needs a preamble/frontmatter. See https://github.com/precice/precice/pull/2385#discussion_r2863478044

I can do the integration with the website after merging.

JulianHarrison-Wirth committed:
hopefully correct implicit cycling

JulianHarrison-Wirth committed:
made a logging helper function to clean up other functions

JulianHarrison-Wirth commented on extras/mock/README.md:
```diff
@@ -0,0 +1,286 @@
+
+## Building
+
+The mock is built as part of the preCICE library.
```
do you mean rebuild the mock after changes or build the mock without any precice components available?

uekerman replied:
To only build the mock, but not libprecice, testprecice, ...
```
make mock
```

JulianHarrison-Wirth committed:
Cleaned up internal state

JulianHarrison-Wirth committed:
started cleanig up readme

JulianHarrison-Wirth committed:
undid changes to unrelated files

JulianHarrison-Wirth committed:
Reverted exact lenght matching for given vector multipliers in scaled buffer mode
use case: giving a single multiplier as a default fallback for ultiple meshes

JulianHarrison-Wirth committed:
Fixed lots of small inconsistencies in the mock

JulianHarrison-Wirth committed:
Add fillReadValues method allowing mapAndReadData to return data as well

JulianHarrison-Wirth committed:
added tests to the mock

JulianHarrison-Wirth committed:
Added Smoke Test and reorganized Tests

JulianHarrison-Wirth committed:
Update extras/mock/preciceMocked.cpp

Co-authored-by: Benjamin Uekermann <benjamin.uekermann@gmail.com>

JulianHarrison-Wirth committed:
Stabilized tests for pre-commit

JulianHarrison-Wirth committed:
changed name to participantName

JulianHarrison-Wirth committed:
Merge branch 'add-mocking-prototype' of https://github.com/JulianHarrison-Wirth/precice into add-mocking-prototype

JulianHarrison-Wirth committed:
adressed remaining comments
mainly removing checks that are redundant with the config checker and removing unused code

JulianHarrison-Wirth committed:
updated the readme to (my) final version

JulianHarrison-Wirth commented on extras/mock/preciceMocked.cpp:
```diff
@@ -0,0 +1,1796 @@
+    if (_impl->mockConfigExists) {
+      std::cout << "---[precice]  Configuring preCICE (mock) with configuration \"" << std::filesystem::absolute(_impl->config).string() << "\" and mock configuration \"" << std::filesystem::absolute(_impl->mockConfigPath).string() << "\"" << std::endl;
+    } else {
+      std::cout << "---[precice]  Configuring preCICE (mock) with configuration \"" << std::filesystem::absolute(_impl->config).string() << "\" and default mock configuration. See the mock-README about how to adjust this." << std::endl;
```
this is the only comment I have not addressed as I don't know where it will end up

MakisH replied:
@JulianHarrison-Wirth I will integrate this repository to the preCICE website, and the `README.md` will be rendered. For this, you will need to add a preamble, similar to https://github.com/precice/precice.github.io/blob/master/content/docs/tooling/tooling-cli.md?plain=1

I agree with `tooling-mock.html` as the target.

Once the preamble is there, you don't need to do anything else - I will take care of the integration on the website side.

JulianHarrison-Wirth committed:
Fixed compilation for bython bindings tests

MakisH commented on extras/mock/tests/tutorials/baselines/courses/B.out:
I need to catch up with the latest state, but I think we should separate here between test cases that are available in this repository, and test cases that are maintained separately. For this reason, I would remove the tutorials and the course outputs from this PR. They are nice showcases for your thesis, but they will immediately become outdated.

Later on, some of these could be moved to the system tests (tutorials repository), if there is an archive.
